internal/resources/gsi.go

```diff
@@ -0,0 +1,509 @@
+package resources
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api"
+	providerschema "github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/schema"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                   = (*GSI)(nil)
+	_ resource.ResourceWithConfigure      = (*GSI)(nil)
+	_ resource.ResourceWithImportState    = (*GSI)(nil)
+	_ resource.ResourceWithValidateConfig = (*GSI)(nil)
+)
+
+// GSI is the GSI resource implementation.
+type GSI struct {
+	*providerschema.Data
+}
+
+func NewGSI() resource.Resource {
+	return &GSI{}
+}
+
+// Metadata returns the query index resource type name.
+func (g *GSI) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_query_indexes"
+}
+
+// Schema defines the schema for the query index resource.
+func (g *GSI) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = GsiSchema()
+}
+
+// Create will create/drop/build/alter a primary or secondary index.
+func (g *GSI) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var ddl string
+
+	// create build index statement.
+	if !plan.BuildIndexes.IsNull() {
+		ddl = fmt.Sprintf(
+			"BUILD INDEX ON `%s`.`%s`.`%s`(%s)",
+			plan.BucketName.ValueString(),
+			plan.ScopeName.ValueString(),
+			plan.CollectionName.ValueString(),
+			listStringValues(plan.BuildIndexes),
+		)
+	} else {
+		// create primary index statement.
+		if plan.IsPrimary.ValueBool() {
+			var indexName string
+			if !plan.IndexName.IsNull() {
+				indexName = plan.IndexName.ValueString()
+			} else {
+				indexName = "#primary"
+			}
+			ddl = fmt.Sprintf(
+				"CREATE PRIMARY INDEX `%s` ON `%s`.`%s`.`%s`  WITH { \"defer_build\": %t,  \"num_replica\": %d }",
+				indexName,
+				plan.BucketName.ValueString(),
+				plan.ScopeName.ValueString(),
+				plan.CollectionName.ValueString(),
+				plan.With.DeferBuild.ValueBool(),
+				plan.With.NumReplica.ValueInt64(),
```

Comment:
From the schema, it seems that NumReplica and DeferBuild are optional fields, and we haven't populated any default values for them. So, what values are we using if the customer doesn't provide them?

Reply:
I don't want client to determine the default.  This should be left to the server.